// jdnconvertiblecalendardateadapter/esm2015/lib/jdnconvertible-calendar-date-adapter.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { DateAdapter, MAT_DATE_FORMATS, MAT_DATE_LOCALE } from '@angular/material';
import { JDNConvertibleCalendarDateAdapter } from './jdnconvertible-calendar-date-adapter';
import { MAT_JDN_DATE_FORMATS } from './jdnconvertible-calendar-date-formats';
export class JDNConvertibleCalendarDateAdapterModule {
}
JDNConvertibleCalendarDateAdapterModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    { provide: DateAdapter, useClass: JDNConvertibleCalendarDateAdapter, deps: [MAT_DATE_LOCALE] }
                ]
            },] }
];
const ɵ0 = MAT_JDN_DATE_FORMATS;
export class MatJDNConvertibleCalendarDateAdapterModule {
}
MatJDNConvertibleCalendarDateAdapterModule.decorators = [
    { type: NgModule, args: [{
                imports: [JDNConvertibleCalendarDateAdapterModule],
                providers: [{ provide: MAT_DATE_FORMATS, useValue: ɵ0 }],
            },] }
];
export { JDNConvertibleCalendarDateAdapter } from './jdnconvertible-calendar-date-adapter';
export { ɵ0 };
