// jdnconvertiblecalendar/dist/src/index.js
"use strict";
/*
 * Copyright © 2018 Lukas Rosenthaler, Rita Gautschy, Benjamin Geer, Ivan Subotic,
 * Tobias Schweizer, André Kilchenmann, and Sepideh Alassi.
 *
 * This file is part of JDNConvertibleCalendar.
 *
 * JDNConvertibleCalendar is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDNConvertibleCalendar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with JDNConvertibleCalendar.  If not, see <http://www.gnu.org/licenses/>.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var JDNConvertibleCalendar_1 = require("./JDNConvertibleCalendar");
var JDNCalendarConversion_1 = require("./JDNCalendarConversion");
exports.JDNConvertibleConversionModule = JDNCalendarConversion_1.JDNConvertibleConversionModule;
var CalendarDate = JDNConvertibleCalendar_1.JDNConvertibleCalendarModule.CalendarDate;
exports.CalendarDate = CalendarDate;
var JDNPeriod = JDNConvertibleCalendar_1.JDNConvertibleCalendarModule.JDNPeriod;
exports.JDNPeriod = JDNPeriod;
var CalendarPeriod = JDNConvertibleCalendar_1.JDNConvertibleCalendarModule.CalendarPeriod;
exports.CalendarPeriod = CalendarPeriod;
var GregorianCalendarDate = JDNConvertibleCalendar_1.JDNConvertibleCalendarModule.GregorianCalendarDate;
exports.GregorianCalendarDate = GregorianCalendarDate;
var JulianCalendarDate = JDNConvertibleCalendar_1.JDNConvertibleCalendarModule.JulianCalendarDate;
exports.JulianCalendarDate = JulianCalendarDate;
var JDNConvertibleCalendar = JDNConvertibleCalendar_1.JDNConvertibleCalendarModule.JDNConvertibleCalendar;
exports.JDNConvertibleCalendar = JDNConvertibleCalendar;
